contracts/adapters/swap/astrovault/src/contract.rs

```diff
@@ -0,0 +1,552 @@
+    let astrovault_router_wasm_msg = match initial_asset {
+        Asset::Native(native_asset) => CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: astrovault_router_contract_address.to_string(),
+            funds: vec![native_asset],
+            msg: to_json_binary(&router::handle_msg::ExecuteMsg::Receive(
+                cw20::Cw20ReceiveMsg {
+                    sender: env.contract.address.to_string(),
+                    amount: amount_in,
+                    msg: to_json_binary(&router_execute_msg)?,
+                },
+            ))?,
+        }),
+        Asset::Cw20(cw20_asset) => CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: cw20_asset.address.to_string(),
+            funds: vec![],
+            msg: to_json_binary(&cw20::Cw20ExecuteMsg::Send {
+                contract: astrovault_router_contract_address.to_string(),
+                amount: cw20_asset.amount,
+                msg: to_json_binary(&router_execute_msg)?,
+            })?,
+        }),
+    };
+
+    let return_denom = match operations.last() {
+        Some(last_op) => last_op.denom_out.clone(),
+        None => return Err(ContractError::SwapOperationsEmpty),
+    };
+
+    // Create the transfer funds back message
+    let transfer_funds_back_msg = WasmMsg::Execute {
+        contract_addr: env.contract.address.to_string(),
+        msg: to_json_binary(&ExecuteMsg::TransferFundsBack {
+            swapper: entry_point_contract_address,
+            return_denom,
+        })?,
+        funds: vec![],
+    };
+
+    Ok(Response::new()
+        .add_attribute("action", "execute_swap")
+        .add_message(astrovault_router_wasm_msg)
+        .add_message(transfer_funds_back_msg)
+        .add_attribute("action", "dispatch_swaps_and_transfer_back"))
+}
+
+pub fn execute_transfer_funds_back(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    swapper: Addr,
+    return_denom: String,
+) -> Result<Response, SkipError> {
+    // Ensure the caller is the contract itself
+    if info.sender != env.contract.address {
+        return Err(SkipError::Unauthorized);
+    }
+
+    // Create the transfer funds back message
+    let transfer_funds_back_msg: CosmosMsg = match deps.api.addr_validate(&return_denom) {
+        Ok(contract_addr) => Asset::new(
+            deps.api,
+            contract_addr.as_str(),
+            Cw20Contract(contract_addr.clone()).balance(&deps.querier, &env.contract.address)?,
+        )
+        .transfer(swapper.as_str()),
+        Err(_) => CosmosMsg::Bank(BankMsg::Send {
+            to_address: swapper.to_string(),
+            amount: deps
+                .querier
+                .query_all_balances(env.contract.address.clone())?,
+        }),
+    };
+
+    let mut msgs = vec![transfer_funds_back_msg];
+
+    // ADDED: Also create the return of cashback funds msg if available
+    if let Some(cashback_addr) = ASTROVAULT_CASHBACK_ADDRESS.may_load(deps.storage)? {
```

Comment:
Question to learn: could we give this back to the user in some way? What is the purpose of sending it back to the cashback_addr here?

Reply:
We are sending it back to the router address so it does not get lost on the adapter contract.

If the user wallet address is available on the adapter we can send it to them, but I'm not seeing this info being sent by the entry-point contract. Is there a way to get this info on the adapter during the execution of the swap?

We can also sent it to the entrypoint contract but then we would need to add some code to the entrypoint contract also to handle the send of cashback to the user during the post-swap-validation.

Reply:
Yeah the entrypoint has access to such a user address, but heard that this would all require adding more logic and likely not worth it for now, thanks for the clarification!